### ISSUE TITLE
iText

### ISSUE THREAD

**Issue by Srilakshman213** [Srilakshman213/upload#1]
> package iText;

import java.awt.Graphics2D;
import java.awt.geom.AffineTransform;
import java.awt.image.BufferedImage;
import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;

import javax.imageio.ImageIO;

import org.springframework.util.StopWatch;

import com.itextpdf.text.DocumentException;
import com.itextpdf.text.pdf.PRStream;
import com.itextpdf.text.pdf.PdfName;
import com.itextpdf.text.pdf.PdfNumber;
import com.itextpdf.text.pdf.PdfObject;
import com.itextpdf.text.pdf.PdfReader;
import com.itextpdf.text.pdf.PdfStamper;
import com.itextpdf.text.pdf.parser.PdfImageObject;

public class iTextDirect {
    public static final float FACTOR = 0.5f;
    public static void main(String[] args) throws DocumentException, IOException {
    	for(int i=1;i<=8;i++) {
    		StopWatch watch = new StopWatch();
            watch.start();
    		String SRC = "C:\\Users\\a720433\\Downloads\\uploads\\Shared\\test"+i+".pdf";
        	String DEST = "C:\\Users\\a720433\\Downloads\\uploads\\Shared\\icomp"+i+".pdf";
    		File file = new File(DEST);
            file.getParentFile().mkdirs();
    		new iTextDirect().manipulatePdf(SRC, DEST);
    		watch.stop();
            System.out.println("Total execution time: "+ watch.getTotalTimeSeconds());
    	}
    }
    public void manipulatePdf(String src, String dest) throws DocumentException, IOException {
        PdfReader reader = new PdfReader(src);
        int n = reader.getXrefSize();
        PdfObject object;
        PRStream stream;
        // Look for image and manipulate image stream
        for (int i = 0; i < n; i++) {
            object = reader.getPdfObject(i);
            if (object == null || !object.isStream())
                continue;
            stream = (PRStream)object;
            if (!PdfName.IMAGE.equals(stream.getAsName(PdfName.SUBTYPE)))
                continue;
            if (!PdfName.DCTDECODE.equals(stream.getAsName(PdfName.FILTER)))
                continue;
            PdfImageObject image = new PdfImageObject(stream);
            BufferedImage bi = image.getBufferedImage();
            if (bi == null)
                continue;
            int width = (int)(bi.getWidth() * FACTOR);
            int height = (int)(bi.getHeight() * FACTOR);
            if (width <= 0 || height <= 0)
                continue;
            BufferedImage img = new BufferedImage(width, height, BufferedImage.TYPE_INT_RGB);
            AffineTransform at = AffineTransform.getScaleInstance(FACTOR, FACTOR);
            Graphics2D g = img.createGraphics();
            g.drawRenderedImage(bi, at);
            ByteArrayOutputStream imgBytes = new ByteArrayOutputStream();
            ImageIO.write(img, "JPG", imgBytes);
            stream.clear();
            stream.setData(imgBytes.toByteArray(), false, PRStream.NO_COMPRESSION);
            stream.put(PdfName.TYPE, PdfName.XOBJECT);
            stream.put(PdfName.SUBTYPE, PdfName.IMAGE);
            stream.put(PdfName.FILTER, PdfName.DCTDECODE);
            stream.put(PdfName.WIDTH, new PdfNumber(width));
            stream.put(PdfName.HEIGHT, new PdfNumber(height));
            stream.put(PdfName.BITSPERCOMPONENT, new PdfNumber(8));
            stream.put(PdfName.COLORSPACE, PdfName.DEVICERGB);
        }
        reader.removeUnusedObjects();
        // Save altered PDF
        PdfStamper stamper = new PdfStamper(reader, new FileOutputStream(dest));
        stamper.setFullCompression();
        stamper.close();
        reader.close();
    }
}

**Comment by Srilakshman213** [comment]
This code is written in Java and uses the iText library to manipulate PDF files. Specifically, it reduces the size of images in a PDF file by applying a scaling factor to them.

The main method of the program loops through a series of PDF files numbered 1 to 8, and for each file, it calls the manipulatePdf method with the source and destination file paths as arguments.

In the manipulatePdf method, the source PDF is read using the PdfReader class. The getXrefSize() method returns the number of indirect objects in the PDF file, and the program iterates over each object to check if it is an image object of type DCTDecode. If so, the program reads the image data and creates a new scaled-down image using the AffineTransform class. The new image is then written back to the PDF file using the PdfStamper class.

The PRStream class is used to represent a stream object in a PDF file, and it is used to manipulate the image data. The PdfImageObject class is used to read the image data from the stream object and create a BufferedImage object. The ImageIO class is used to write the new image data to a byte array.

The PdfObject class is used to represent PDF objects, and the program checks if an object is an image by checking its subtype. If the subtype is "Image" and the filter is "DCTDecode," the program proceeds to manipulate the image.

The PdfName class is used to represent PDF names, such as "Subtype" and "Filter," and the program uses it to set the values of various properties of the image object. For example, it sets the image width and height, the number of bits per component, the color space, and the filter type.

Finally, the PdfStamper class is used to save the modified PDF file to the destination file path. The setFullCompression() method is called to apply full compression to the PDF file, reducing its size further.

In summary, this code reduces the size of images in a PDF file by scaling them down and compressing the file. It uses the iText library to read and manipulate PDF files and relies on various classes and methods to achieve its objective.

**Comment by Srilakshman213** [comment]
In conclusion, PDF compression is an effective way to reduce the file size of PDF documents, making them easier to share, store, and transmit over the internet. By compressing PDFs, users can save disk space, reduce upload and download times, and make documents more accessible to users with limited bandwidth or storage capacity. Additionally, PDF compression can also improve the overall efficiency of document workflows, particularly in industries such as healthcare, legal, and finance. However, it is important to note that not all compression methods are equal, and users should carefully consider their options before choosing a compression tool or technique. With the right approach, PDF compression can be a powerful tool for streamlining document management and enhancing productivity.

**Comment by Srilakshman213** [comment]
ByteArrayOutputStream imgBytes = new ByteArrayOutputStream();
ImageIO.write(img, "JPG", imgBytes);

Try other formats

stream.setData(imgBytes.toByteArray(), false, PRStream.NO_COMPRESSION);

Try compression here
